core/src/test/java/org/apache/druid/math/expr/ParserTest.java

```diff
@@ -172,6 +173,43 @@ public void testMixed()
     validateFlatten("min(1, max(3, 4))", "(min [1, (max [3, 4])])", "1");
   }
 
+  @Test
+  public void testBitwiseOps()
+  {
+    validateFlatten("3 & 1", "(& 3 1)", "1");
+    validateFlatten("3 & 1", "(& 3 1)", "1");
+    validateFlatten("2 & 1", "(& 2 1)", "0");
+    validateFlatten("3 | 1", "(| 3 1)", "3");
+    validateFlatten("2 | 1", "(| 2 1)", "3");
+    validateFlatten("(~1) & 7", "(& ~1 7)", "6");
+
+    validateFlatten("'2' & '1'", "(& 2 1)", "0");
+    validateFlatten("'3' | '1'", "(| 3 1)", "3");
+    validateFlatten("(~'1') & 7", "(& ~1 7)", "6");
+
+    validateFlatten("'notanumber' & '1'", "(& notanumber 1)", null);
+    validateFlatten("'3' | 'notanumber'", "(| 3 notanumber)", null);
+    validateFlatten("~'notanumber'", "~notanumber", null);
+    validateFlatten("(~'notanumber') & '7'", "(& ~notanumber 7)", null);
+    validateFlatten("(~'notanumber') | '7'", "(| ~notanumber 7)", null);
```

Comment:
could you add some tests with nulls for the bitwise operators

Reply:
Hmm, this actually brings up something worth thinking about actually, and I'm not exactly certain what the most correct behavior is to have here. 

A test (at least here) would use a `null` constant, which `BinaryEvalOpExprBase` currently is _not_ short circuited to return `null` unless `NullHandling.sqlCompatible()` is set.

The "default" evaluator in `BinaryEvalOpExprBase` is `evalDoubles` if none of the other conditions match, and the other evaluators (`evalString` and `evalLong`) both require left and right expressions to be the same type, which the `null` constant makes not be true unless the other expression is `ExprType.STRING`.

This means tests using a `null` constant and a number on the other side end up in `evalDoubles`, which for these bitwise expressions will
```
throw new IllegalArgumentException("unsupported type " + ExprType.DOUBLE);
```
which seems sort of funny, and not really correct.

I'm not really certain that `evalDouble` being the default makes sense, but even if it does, i'm not certain what the correct behavior in default mode if a null constant is involved (there would never be a null number from a column in default mode).

I think it really comes down to what should be the behavior in default mode of the `null` constant when used in a number expression. The `'notanumber'` string case evaluates to null eventually, even in default mode, while `null` does not, which seems inconsistent.

Reply:
```'notanumber' & 1``` actually has the same problem of ending up as `evalDouble` and exploding, so it's not _strictly_ the null constant, but `'notanumber'` ends up as a `null` so it's the same-ish problem.

Reply:
Maybe `BinaryEvalOpExprBase` should have a special case for `null` literals?